# **Tarea 2 Módulo 3**

## Alumno: Nicolás Bene

---

Vamos a analizar el perfil de Rosalía (@rosalia) en Twitter. 

Las tareas que debes realizar y las preguntas a las que debes dar respuesta en esta actividad son la siguientes:

1. Autentícate en twitter. Pon en notebook todo el código necesario y deja con “XXXXXX” las credenciales (bearer_token) para que el profesor se pueda autenticar con sus propias credenciales y ejecutar el resto del notebook. 
1. ¿Cuál es el id de @rosalia en twitter?
1. Lista los nombres de los usuarios a los que @rosalia está siguiendo.
1. ¿Cuál de ellos tiene más seguidores?
1. Utiliza la get_users_tweets() para obtener el json completo con los cinco últimos tweets de Rosalia. Muestra por pantalla el día, la hora de creación y el texto.
1. Recupera los cinco primeros tweets, a partir de la SteamingClient, desde de la entidad del país/región/ciudad de tu elección y que no sean retweets. Muestra por pantalla el id del usuario que realiza el tweet, la fecha, la hora de creación y el texto. 

<br>

Primero importo las librerías

In [25]:
%%capture

#Instalo el upgrade de tweepy (solo para colab, no es necesario en Jupyter). 
#Se utilizo el comando mágico capture al inicio de la celda para que no salga el output de la instalación.
if 'google.colab' in str(get_ipython()):
    !pip install tweepy --upgrade

import tweepy
import pandas as pd

## 1) Autentícate en twitter. Pon en notebook todo el código necesario y deja con “XXXXXX” las credenciales (bearer_token) para que el profesor se pueda autenticar con sus propias credenciales y ejecutar el resto del notebook. 

Creo las credenciales para posteriormente autenticarme y conectarme a la API de Twitter. Para ello primero creo una variable con mi bearer token.

In [26]:
bearer_token = 'AAAAAAAAAAAAAAAAAAAAAGhfZwEAAAAApP%2BYZkOjG7iDq1qiA9clljU5O3A%3DdWntbbi64GvVGo6oZe1VU1tvowTORQdie9AsexHwsDq60rqyYe'

A continuación me autentico.

In [27]:
client = tweepy.Client(bearer_token = bearer_token, wait_on_rate_limit = True)

## 2) ¿Cuál es el id de @rosalia en twitter?

Para ver el id de la usuaria solicitada, primero creo una variable con el nombre de la misma y obtengo sus datos a través de la API, especificando los `user_fields`.

In [28]:
USER_NAME = "rosalia"

#Obtengo los datos del usuario rosalia
user = client.get_user(username = USER_NAME, 
                       user_fields = ["created_at","description","entities","id","location","name","pinned_tweet_id"
                       ,"profile_image_url","protected","public_metrics","url","username","verified","withheld"])

Obtenido así el usuario, me fijo el id analizando dicho atributo del objeto user.

In [29]:
user.data.id

1728956976

El id de @rosalia es, entonces, 1728956976.

## 3) Lista los nombres de los usuarios a los que @rosalia está siguiendo.

Primero me fijo la cantidad de usuarios a los que @rosalia sigue.

In [30]:
user.data.public_metrics["following_count"]

54

Son 54, por lo tanto se pueden listarlos fácilmente por pantalla.

Para verlos uso el comando `get_users_following`, ya que el mismo muestra las personas a las que sigue @rosalia. Para listarlos a todos  se puede utilizar un objeto `Paginator`, y dentro de cada página me quedo con el atributo `username` que es el nombre del usuario. Cada nombre de usuario lo voy asignando a una lista creada previamente, llamada following.

Posteriormente controlaré que la cantidad de elementos de esta lista coincida con el conteo realizado anteriormente.

In [31]:
#Creo la lista en la que luego almacenaré los nombres de usuarios
following = []

#Realizo el loop con el que se llenará la lista anterior
for page in tweepy.Paginator(client.get_users_following, user.data.id, max_results=100, limit = 14):
  for response in page.data:
    following.append(response.data["username"])

El loop anterior creó la lista following, donde se guardó el nombre de los usuarios a los que sige @rosalia. Me fijo el número de elementos de dicha lista para cerciorarme que da igual que el obtenido usando `user.data.public_metrics["following_count"]`.

In [32]:
len(following)

54

El número de elementos es igual que el conteo que se hizo. Resta listar los nombres de los 54 usuarios a los que Rosalia sigue:

In [33]:
for i in following:
    print(i)

NoeliaRMontes
jose272011
neggata
jooleesah
rociorotten
caesarenas
katbeee
motovenus
zanelowe
dantijerina
RollingStoneUK
carolineplz
ivangb00
iamcardib
Tainy
rossydpalma
officialDannyT
DONOMAR
OTHERtone
brokemogul
ChadHugo
Pharrell
TheNeptunes
babykeem
Harry_Styles
tokischa_
secoweedcodein
kaydyCain
dragon___loco
caetanoveloso
0PN
jimmyfallon
MuraiHiro
MayaCPopa
historiayarte_
yokoono
elonmusk
rauwalejandro
MissyElliott
YouTubeJapan
jamesblake
CharmLadonna
arca1000000
Camaron_Fans
ptobella
playboicarti
caroshawmusic
KylieJenner
ChloeBailey
rojuuaka
RodoredaMerce
leticia__sala
michaeluzowuru
daikyri_


## 4) ¿Cuál de ellos tiene más seguidores?

Para contestar a esta pregunta, lo primero que hago es un loop para crear una lista con la cantidad followers de cada uno de los 54 usuarios a los que sigue Rosalia. Para ello en el loop obtengo cada uno de los usuarios de la lista following con `client.get_user`, y sobre los mismos aplico `data.public_metrics["followers_count"]` para obtener el número de usuarios que los siguen.

In [35]:
#Creo la lista en la que luego almacenaré el conteo de los followers de los usuarios a los que sigue rosalia
f_counts = []

#Realizo el loop que llena la lista anterior
for i in following:
    f_counts.append(
        client.get_user(username = i, 
                        user_fields = ["public_metrics"])
            .data.public_metrics["followers_count"]
        )

Luego paso las dos listas creadas (la de usuarios a los que rosalia sigue y la del conteo de los followers de los mismos) a un dataframe.

In [50]:
#Creo el dataframe
following_df = pd.DataFrame({"user" : following, "followers_counts": f_counts})

#Lo observo y lo ordeno en forma descendente para ver cual usuario tiene más followers
following_df.sort_values("followers_counts", ascending= False)

user  followers_counts
36        elonmusk          77604248
31     jimmyfallon          51455441
47     KylieJenner          39697776
24    Harry_Styles          37434555
13       iamcardib          22118077
21        Pharrell          10659795
38    MissyElliott           6967457
17         DONOMAR           5279803
35         yokoono           4609237
45    playboicarti           2230327
37   rauwalejandro           1845017
48     ChloeBailey           1148487
29   caetanoveloso           1027211
8         zanelowe            676643
16  officialDannyT            469434
39    YouTubeJapan            414267
40      jamesblake            339635
23        babykeem            206295
26  secoweedcodein            160621
27       kaydyCain            135238
25       tokischa_            122179
42     arca1000000            118787
14           Tainy            105417
34  historiayarte_             96800
11     carolineplz             81798
49        rojuuaka             76560
15     rossydpalma             67924
30             0PN             67648
19      brokemogul             60681
1       jose272011             39026
20        ChadHugo             35122
53        daikyri_             34785
50   RodoredaMerce             31308
32       MuraiHiro             31191
43    Camaron_Fans             24278
33       MayaCPopa             21191
2          neggata             20800
22     TheNeptunes             19219
52  michaeluzowuru             10016
46   caroshawmusic              8074
41    CharmLadonna              7454
51   leticia__sala              6374
6          katbeee              4182
0    NoeliaRMontes              3558
10  RollingStoneUK              3470
3        jooleesah              3443
18       OTHERtone              1857
5       caesarenas              1585
4      rociorotten              1471
44        ptobella              1178
12        ivangb00               557
9      dantijerina               345
28   dragon___loco               280
7        motovenus                46

Dentro de los 54 twitteros, Elon Musk es el que tiene más seguidores, con más de 77 millones.

## 5) Utiliza la get_users_tweets() para obtener el json completo con los cinco últimos tweets de Rosalia. Muestra por pantalla el día, la hora de creación y el texto. ##

In [43]:
#Busco los últimos 5 tweets de Rosalia, y establezco que incluya el campo created_at, ya que el texto lo trae por defecto
tweets = client.get_users_tweets(id = user.data.id, tweet_fields = ["created_at"], max_results =5)

#Imprimo los datos solicitados de los tweets
conteo = 1
for tweet in tweets.data:
    print("Tweet N°",conteo,"\n",
          "Fecha y hora de creación:", tweet.created_at,  "\n",
          tweet.text, "\n")
    conteo+=1

Tweet N° 1 
 Fecha y hora de creación: 2022-03-13 00:06:10+00:00 
 RT @RosaliaSources: En 4 horas y 30 minutos comenzará #SNL, donde @Rosalia realizará 2 actuaciones.

Link para verlo en directo: 
https://t… 

Tweet N° 2 
 Fecha y hora de creación: 2022-03-13 00:00:46+00:00 
 RT @RosaliaInfoES: ROSALÍA — “Saturday Night Live”

 ➤ Sábado, 12 de marzo:
9:30 PM 🇨🇷🇸🇻🇬🇹🇭🇳🇳🇮
10:30 PM 🇪🇨🇵🇪🇲🇽🇵🇦🇨🇴
11:30 PM 🇻🇪🇩🇴🇧🇴🇨🇺🇨🇱🇵🇾🇵🇷… 

Tweet N° 3 
 Fecha y hora de creación: 2022-03-12 04:38:14+00:00 
 RT @johandls360: NO PUEDO ESPERARRRRrrrrrr @rosalia #MOTOMAMI https://t.co/rN1aC2ibc3 

Tweet N° 4 
 Fecha y hora de creación: 2022-03-12 00:06:26+00:00 
 @tokischa_ @jimmyfallon Eso tUUUUU todo el rato toKi💓 

Tweet N° 5 
 Fecha y hora de creación: 2022-03-11 21:07:02+00:00 
 Aunque
14/20:

El casco es el mejor bolso de unA motomami 



## 6) Recupera los cinco primeros tweets, a partir de la SteamingClient, desde de la entidad del país/región/ciudad de tu elección y que no sean retweets. Muestra por pantalla el id del usuario que realiza el tweet, la fecha, la hora de creación y el texto.

In [48]:
# Declaro la clase listener que va a instanciar tweepy.StreamingClient 
class listener(tweepy.StreamingClient):
    # Creo un contador de tweets capturados
    def on_connect(self):
        self.counter = 1
        self.limit = 6
    # Evento que se ejecuta con cada tweet
    def on_tweet(self, tweet):
        print("Tweet N°",self.counter,"\n",  
              "id del usuario: ", tweet.id, "\n",
              "Fecha y hora de creación:", tweet.created_at,"\n",
             tweet.text,"\n"
              )
        self.counter += 1
        if self.counter < self.limit:
            return True
        else:
            self.disconnect()
    def on_error(self, tweet):
        print(tweet)

#Declaro la instancia de tweepy.StreamingClient
listen = listener(bearer_token,wait_on_rate_limit=True)

#Elimino las reglas previas
printer_rules = listen.get_rules()
if printer_rules.data is not None:
  listen.delete_rules([str(i.id) for i in listen.get_rules().data])

#Declaro las nuevas reglas, donde mostraré tweets de Uruguay que no sean retweets
listen.add_rules(tweepy.StreamRule('entity:"Uruguay" -is:retweet'))

# Filtro e indico los campos a recuperar. No menciono explicitamente el id ni el texto porque lo muestra por default.
listen.filter(tweet_fields=["created_at"])

Tweet N° 1 
 id del usuario:  1503039733102002178 
 Fecha y hora de creación: 2022-03-13 16:06:12+00:00 
 El círculo está tirando más para el lado de Uruguay, pero yo me cruzo, donde ellos vayan yo voy!!!! https://t.co/EM4o4G9Bef 

Tweet N° 2 
 id del usuario:  1503039765326798849 
 Fecha y hora de creación: 2022-03-13 16:06:19+00:00 
 @NetoCavagneri Aah, Bengoechea entonces no sabe nada, campeón en Uruguay y Perú de pura suerte, ah ya. 

Tweet N° 3 
 id del usuario:  1503039820238450688 
 Fecha y hora de creación: 2022-03-13 16:06:32+00:00 
 @afigue2010 @Chasquetti @marianapomies Está campaña tiene muchas cosas novedosas, lamentablemente algunas bordean la inconstitucionalidad y lo políticamente ilógico! La tradición democrática del uruguay merece los mejores estándares republicanos 

Tweet N° 4 
 id del usuario:  1503039838525833220 
 Fecha y hora de creación: 2022-03-13 16:06:37+00:00 
 @ECavaniOfficial @ManUtd @premierleague Start of the warm up games for the next Uruguay internati

Stream connection closed by Twitter


Tweet N° 5 
 id del usuario:  1503039964296134658 
 Fecha y hora de creación: 2022-03-13 16:07:07+00:00 
 La policía lo tiene que custodiar por 180 días, 24hs, y lo acompañan a COMPRAR DROGA.... URUGUAY NOMÁÁÁÁ https://t.co/rHhUY9cyv0 

